core/commands/files.go

```diff
@@ -423,60 +411,53 @@ Examples:
 		cmdkit.BoolOption(longOptionName, "Use long listing format."),
 		cmdkit.BoolOption(dontSortOptionName, "Do not sort; list entries in directory order."),
 	},
-	Run: func(req oldcmds.Request, res oldcmds.Response) {
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
 		var arg string
 
-		if len(req.Arguments()) == 0 {
+		if len(req.Arguments) == 0 {
 			arg = "/"
 		} else {
-			arg = req.Arguments()[0]
+			arg = req.Arguments[0]
 		}
 
 		path, err := checkPath(arg)
 		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+			return err
 		}
 
-		nd, err := req.InvocContext().GetNode()
+		nd, err := cmdenv.GetNode(env)
 		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+			return err
 		}
 
 		fsn, err := mfs.Lookup(nd.FilesRoot, path)
 		if err != nil {
-			res.SetError(err, cmdkit.ErrNormal)
-			return
+			return err
 		}
 
-		long, _, _ := req.Option(longOptionName).Bool()
+		long, _ := req.Options[longOptionName].(bool)
 
 		switch fsn := fsn.(type) {
 		case *mfs.Directory:
 			if !long {
 				var output []mfs.NodeListing
-				names, err := fsn.ListNames(req.Context())
+				names, err := fsn.ListNames(req.Context)
 				if err != nil {
-					res.SetError(err, cmdkit.ErrNormal)
-					return
+					return err
 				}
 
 				for _, name := range names {
 					output = append(output, mfs.NodeListing{
 						Name: name,
 					})
 				}
-				res.SetOutput(&filesLsOutput{output})
-			} else {
-				listing, err := fsn.List(req.Context())
-				if err != nil {
-					res.SetError(err, cmdkit.ErrNormal)
-					return
-				}
-				res.SetOutput(&filesLsOutput{listing})
+				return res.Emit(&filesLsOutput{output})
```

Comment:
Should use `cmds.EmitOnce` (because http headers)

Reply:
> // EmitOnce is a helper that emits a value wrapped in Single, to signal that this will be the only value sent.

This is the `EmitOnce` comment.

@magik6k  I'm not familiar with `go-ipfs-cmds`, so I want to check with you. Should we replace all `res.Emit()` with `cmds.EmitOnce`?

Reply:
We want to use it only in places which return a single value. For channels and readers we want to keep using `res.Emit`. The header logic which is affected by this is in https://github.com/ipfs/go-ipfs-cmds/blob/master/http/responseemitter.go#L266

Reply:
@magik6k Got it, thx.